rust/candid/src/types/subtype.rs

```diff
@@ -0,0 +1,83 @@
+use super::internal::{find_type, Field, Label, Type};
+use crate::parser::typing::TypeEnv;
+use std::collections::{HashMap, HashSet};
+
+/// Check if t1 <: t2
+pub fn subtype(env: &TypeEnv, t1: &Type, t2: &Type) -> bool {
+    use Type::*;
+    if t1 == t2 {
+        return true;
+    }
+    match (t1, t2) {
+        (_, Var(id)) => subtype(env, t1, env.rec_find_type(id).unwrap()),
+        (Var(id), _) => subtype(env, env.rec_find_type(id).unwrap(), t2),
+        (_, Knot(id)) => subtype(env, t1, &find_type(id).unwrap()),
+        (Knot(id), _) => subtype(env, &find_type(id).unwrap(), t2),
+        (_, Reserved) => true,
+        (Empty, _) => true,
+        (Nat, Int) => true,
+        (Vec(ty1), Vec(ty2)) => subtype(env, ty1, ty2),
+        (Null, Opt(_)) => true,
+        (Opt(ty1), Opt(ty2)) if subtype(env, ty1, ty2) => true,
+        (t1, Opt(ty2)) if subtype(env, t1, ty2) && !subtype(env, &Null, ty2) => true,
+        (Opt(ty1), Opt(_ty2)) if !subtype(env, ty1, t2) => {
+            eprintln!("{} <: {} via special opt rule", t1, t2);
+            true
+        }
+        (Opt(ty1), Opt(ty2)) if !subtype(env, ty1, ty2) && !subtype(env, &Null, ty1) => {
+            eprintln!("{} <: {} via special opt rule", t1, t2);
+            true
+        }
```

Comment:
You can simplify these, and simply use `t < opt t'` for all types `t` and `t'`. The spec lists more complicated rules, but really just for educational purposes; that single rule would suffice (also cf. the definition in the Coq proof)

Reply:
From the rules in the spec, `opt opt int <: opt int` doesn't hold.

Reply:
By this rule it should?
```
not (<datatype> <: <datatype'>)
---------------------------------
opt <datatype> <: opt <datatype'>
```

Really, the intention is that every type is a subtype of `opt t`, else using that for new optional fields in records doesn't works. I'd the rules in the spec are not exhaustive here, that's a bug in the spec.

Reply:
It's a typo then. The rule in the spec says:

```
not (<datatype> <: opt <datatype'>)
---------------------------------
opt <datatype> <: opt <datatype'>
```

Reply:
Hmm, indeed, I must have copied from a wrong version. But note that the prose says

> Finally, in order to maintain transitivity of subtyping, two unusual rules allow, in fact, any type to be regarded as a subtype of an option.

so `opt opt int <: opt int` _must_ hold, and if it doesn't, the rules are still buggy.

I see this rule in the spec:
```
not (null <: <datatype>)
not (<datatype> <: <datatype'>)
---------------------------------
opt <datatype> <: opt <datatype'>
```
but I think it should be
```
not (null <: <datatype'>)
not (<datatype> <: <datatype'>)
---------------------------------
opt <datatype> <: opt <datatype'>
```
to complement the rule
```
not (null <: <datatype'>)
<datatype> <: <datatype'>
-----------------------------
<datatype> <: opt <datatype'>
```